pkg/autodiscovery/autoconfig.go

```diff
@@ -269,20 +280,94 @@ func (ac *AutoConfig) resolve(config integration.Config) []integration.Config {
 	return configs
 }
 
-// AddListener adds a service listener to AutoConfig.
-func (ac *AutoConfig) AddListener(listener listeners.ServiceListener) {
+// AddListeners tries to initialise the listeners listed in the given configs. A first
+// try is done synchronously. If a listener fails with a ErrWillRetry, the initialization
+// will be re-triggered later until success or ErrPermaFail.
+func (ac *AutoConfig) AddListeners(listenerConfigs []config.Listeners) {
+	ac.addListenerCandidates(listenerConfigs)
+	remaining := ac.initListenerCandidates()
+	if remaining == false {
+		return
+	}
+
+	// Start the retry logic if we have remaining candidates and it is not already running
+	ac.m.Lock()
+	defer ac.m.Unlock()
+	if ac.listenerRetryStop == nil {
+		ac.listenerRetryStop = make(chan struct{})
+		go ac.retryListenerCandidates()
+	}
+}
+
+func (ac *AutoConfig) addListenerCandidates(listenerConfigs []config.Listeners) {
 	ac.m.Lock()
 	defer ac.m.Unlock()
 
-	for _, l := range ac.listeners {
-		if l == listener {
-			log.Warnf("Listener %s was already added, skipping...", listener)
-			return
+	for _, c := range listenerConfigs {
+		factory, ok := listeners.ServiceListenerFactories[c.Name]
+		if !ok {
+			// Factory has not been registered.
+			log.Warnf("Listener %s was not registered", c.Name)
+			continue
+		}
+		log.Debugf("Listener %s was registered", c.Name)
+		ac.listenerCandidates[c.Name] = factory
+	}
+}
+
+func (ac *AutoConfig) initListenerCandidates() bool {
+	ac.m.Lock()
+	defer ac.m.Unlock()
+
+	for name, factory := range ac.listenerCandidates {
+		listener, err := factory()
+		switch {
+		case err == nil:
+			// Init successful, let's start listening
+			log.Infof("%s listener successfully started", name)
+			ac.listeners = append(ac.listeners, listener)
+			listener.Listen(ac.configResolver.newService, ac.configResolver.delService)
+			delete(ac.listenerCandidates, name)
+		case retry.IsErrWillRetry(err):
+			// Log an info and keep in candidates
+			log.Infof("%s listener cannot start, will retry: %s", name, err)
+		default:
+			// Log an error and remove from candidates
+			log.Errorf("%s listener cannot start: %s", name, err)
+			delete(ac.listenerCandidates, name)
 		}
 	}
 
-	ac.listeners = append(ac.listeners, listener)
-	listener.Listen(ac.configResolver.newService, ac.configResolver.delService)
+	if len(ac.listenerCandidates) > 0 {
```

Comment:
you can shortcut to `return len(ac.listenerCandidates) > 0` classic `if bool { return bool }` trick 😁